File: cargo/crates/numbers-utils/tests/operations/factor_utils_tests.rs
use chrono::Local;
use colored::Colorize;
use num_bigint::BigUint;

use numbers_utils::operations::factor_utils::{
    factor, factor_random, prime_factor, prime_factor_random,
};

/// # `format_error_message`
///
/// This function formats the error message for the test cases.
///
/// __Arguments:__
///
/// * `expected: &Vec<BigUint>` - The expected output.
/// * `actual: &Vec<BigUint>` - The actual output.
///
/// __Return:__
///
/// * `colored::ColoredString` - The formatted error message.
fn format_error_message(expected: &Vec<BigUint>, actual: &Vec<BigUint>) -> colored::ColoredString {
    format!(
        "\n\nTest run at: {}\nExpected : {:?}\nActual   : {:?}\n",
        Local::now().to_rfc3339(),
        expected,
        actual
    )
    .red()
}

/// # `factor_test`
///
/// This function tests the `factor` function.
#[test]
fn factor_test() {
    let number = BigUint::from(60u32);
    let expected_output = vec![
        BigUint::from(1u32),
        BigUint::from(2u32),
        BigUint::from(3u32),
        BigUint::from(4u32),
        BigUint::from(5u32),
        BigUint::from(6u32),
        BigUint::from(10u32),
        BigUint::from(12u32),
        BigUint::from(15u32),
        BigUint::from(20u32),
        BigUint::from(30u32),
        BigUint::from(60u32),
    ];
    let actual_output = factor(number);
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(actual_output, expected_output, "{}", error_message);
}

/// # `factor_random_test`
///
/// This function tests the `factor_random` function.
#[test]
fn factor_random_test() {
    let result = factor_random();
    let start_string = "Factor[";
    let error_message = format!("\nExpected: {}\n     Got: {}\n", start_string, result).red();
    assert!(result.starts_with(start_string), "{}", error_message,);
    assert!(result.contains(" = "));
}

/// # `prime_factor_test`
///
/// This function tests the `prime_factor` function.
#[test]
fn prime_factor_test() {
    let number = BigUint::from(60u32);
    let expected_output = vec![
        BigUint::from(2u32),
        BigUint::from(2u32),
        BigUint::from(3u32),
        BigUint::from(5u32),
    ];
    let actual_output = prime_factor(number);
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(actual_output, expected_output, "{:?}", error_message);
}

/// # `prime_factor_random_test`
///
/// This function tests the `prime_factor_random` function.
#[test]
fn prime_factor_random_test() {
    let result = prime_factor_random();
    let start_string = "PFactor[";
    let error_message = format!("\nExpected: {}\n     Got: {}\n", start_string, result).red();
    assert!(result.starts_with(start_string), "{}", error_message,);
    assert!(result.contains(" = "));
}

File: kotlin/subprojects/numbers-utils/src/test/kotlin/com/prog/numbers/utils/operations/FactorUtilsTests.kt
package com.prog.numbers.utils.operations

import org.apache.commons.lang3.time.DateFormatUtils
import java.math.BigInteger
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class FactorUtilsTests {
    /**
     * # `formatErrorMessage`
     *
     * This function formats the error message for the test.
     *
     * __Arguments:__
     *
     * * `expected: List<BigInteger>` - The expected output.
     * * `actual: List<BigInteger>` - The actual output.
     *
     * __Return:__
     *
     * * `String` - The formatted error message.
     */
    private fun formatErrorMessage(
        expected: List<BigInteger>,
        actual: List<BigInteger>,
    ): String {
        val now = System.currentTimeMillis()
        val formattedDate = DateFormatUtils.format(now, "yyyy-MM-dd'T'HH:mm:ss")
        return """
            |${"\u001B[31m"}
            |Test run at: $formattedDate
            |Expected : $expected
            |  Actual : $actual
            |
            |${"\u001B[00m"}
            """.trimMargin()
    }

    /**
     * # `factor test`
     *
     * This function tests the `factor` function.
     */
    @Test
    fun `factor test`() {
        val number = BigInteger.valueOf(60)
        val expectedOutput =
            listOf(
                BigInteger.valueOf(1),
                BigInteger.valueOf(2),
                BigInteger.valueOf(3),
                BigInteger.valueOf(4),
                BigInteger.valueOf(5),
                BigInteger.valueOf(6),
                BigInteger.valueOf(10),
                BigInteger.valueOf(12),
                BigInteger.valueOf(15),
                BigInteger.valueOf(20),
                BigInteger.valueOf(30),
                BigInteger.valueOf(60),
            )
        val actualOutput = FactorUtils.factor(number)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }

    /**
     * # `factor random test`
     *
     * This function tests the `factorRandom` function.
     */
    @Test
    fun `factorial random test`() {
        val result = FactorUtils.factorRandom()
        val startString = "Factor["
        val errorMessage =
            """
            |${"\u001B[31m"}
            |Expected: $startString
            |     Got: $result${"\u001B[00m"}
            """.trimMargin()
        assertTrue(result.startsWith(startString), errorMessage)
        assertTrue(result.contains(" = "))
    }

    /**
     * # `prime factor test`
     *
     * This function tests the `primeFactor` function.
     */
    @Test
    fun `prime factor test`() {
        val number = BigInteger.valueOf(60)
        val expectedOutput =
            listOf(
                BigInteger.valueOf(2),
                BigInteger.valueOf(2),
                BigInteger.valueOf(3),
                BigInteger.valueOf(5),
            )
        val actualOutput = FactorUtils.primeFactor(number)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }

    /**
     * # `prime factorial random test`
     *
     * This function tests the `primeFactorRandom` function.
     */
    @Test
    fun `prime factorial random test`() {
        val result = FactorUtils.primeFactorRandom()
        val startString = "PFactor["
        val errorMessage =
            """
            |${"\u001B[31m"}
            |Expected: $startString
            |     Got: $result${"\u001B[00m"}
            """.trimMargin()
        assertTrue(result.startsWith(startString), errorMessage)
        assertTrue(result.contains(" = "))
    }
}
